feat(vpc): add check if BastionInstanceProfile is provided and skip creating Role

mkilp opened the pull request:
Hi everyone,

similar to #6269 `createRole` is often restricted in bigger organizations but I would still like to use the bastion feature of vpcs.

This PR adds a check if a `iam.InstanceProfile` is provided and uses that instead of creating a new role and profile.

LMK if you need any changes. Appreciate all input.

Thanks,

Marv

mkilp committed:
feat(vpc): add check if BastionInstanceProfile is provided and skip creating Role

mkilp commented:
@vimtor do you think I could get a review on this? We would love to use the bastion functionality of sst but IAM permissions are blocking us. This PR checks for transform but if you prefer I can also make it a component level setting aka

```ts
new sst.aws.Vpc("MyVPC", {
  bastion: {
    enabled: true,
    instanceProfile: <exampleArn>,
  }
});
```

vimtor commented:
@mkilp i will try to review both of these today!

vimtor commented:
@mkilp after reviewing the code i much rather prefer:

```ts
new sst.aws.Vpc("MyVPC", {
  bastion: {
    enabled: true, // probably not needed
    instanceProfile: <exampleArn>,
  }
});
```

mkilp committed:
feat(vpc): enhance bastion configuration with instance profile support

mkilp commented:
@vimtor change is done. Added a normalize function to the bastion to be in line with the rest of SST. I also then defaulted "enabled" to true and modified parts of the app where the bastion component is checked!

LMK if you need changes.

vimtor commented on platform/src/components/aws/vpc.ts:
```diff
@@ -1239,33 +1270,54 @@ export class Vpc extends Component implements Link.Linkable {
-            {
-              assumeRolePolicy: iam.getPolicyDocumentOutput({
-                statements: [
+          const instanceProfile = bastion.instanceProfileName
```
it's better to follow this pattern in pulumi:

```ts
return output(bastions.instanceProfileName).apply(instanceProfileName => {
  if (instanceProfileName) {
    return iam.InstanceProfile.get();
  }
  
  return new iam.InstanceProfile();
});
```

it avoids many bugs

mkilp replied:
Should we apply this everywhere for iam getters? In fargate.ts for example we do

```ts
if (args.taskRole)
    return iam.Role.get(`${name}TaskRole`, args.taskRole, {}, { parent });
``` 

(https://github.com/anomalyco/sst/blob/dev/platform/src/components/aws/fargate.ts#L871)
which is why i assumed a direct getter would be fine.
Happy to rewrite it to the output pattern tho

vimtor committed:
refactor(vpc): use output().apply() for bastion instance profile

vimtor reviewed (approved):
thanks for your contribution @mkilp 

<details>

<summary>just applied the minor fixed suggested but everything else is currently working</summary>

<img width="1624" height="696" alt="Image" src="https://github.com/user-attachments/assets/2115ed60-3d1c-4c7d-867c-d5fe21888aad" />

<img width="2444" height="1468" alt="Image" src="https://github.com/user-attachments/assets/ac64af9c-24e9-417f-b3a4-3cd4e5e3f194" />

</details>
